### PR TITLE
CIRC-371: Refuse new request for patron who already placed a request on a copy of the same title

### PR DESCRIPTION
**Pull request by mtraneis**
## Purpose
It was clarified that a patron cannot place multiple requests on various copies of the same title.  In this implementation we are going to validate the requester against the requests in the items' request queues.

## Approach

- Added a validateRequester method to check the requesterId against the requesterIds in the requests of the request queues. If any existing request has matching requesterId, then fail the instance-level request with the message "This requester already has an open request for a copy of this title".
- Previously getting both request queues and loans were done in parallel because one wasn't depending on the other. Now there's a need to validate the requester against all instance's items' request queues, so this check has to be done first, If the validation passes then the step to get the loans would come after. Therefore once-again all the calls could be chained together and no need to break up the futures chain. 

#### TODOS and Open Questions
- Should this validation be done on loans as well? We don't want to allow the same requester to checkout a copy and subsequently place requests on other copies of the instance?

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 1]
```diff
@@ -364,33 +358,63 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
...
+          .stream()
+          .filter(request -> request.isOpen()
+                           && Objects.equals(request.asJson().getString(REQUESTER_ID_FIELD),
+                                             requesterId.toString())
+          ).findFirst();
+
+        if (matchingRequest.isPresent()) {
+            Map<String, String> parameters = new HashMap<>();
+            parameters.put(REQUESTER_ID_FIELD, requesterId.toString());
+            parameters.put(ITEM_ID_FIELD, matchingRequest.get().asJson().getString(ITEM_ID_FIELD));
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Can we easily also include the instance ID? 

I prefer that the validation messages refer to input parameters where possible (inspired by the guidance on JSON API, though we don’t apply the level of guidelines it suggests).

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 2]
```diff
@@ -364,33 +358,63 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
...
+            String message = "This requester already has an open request for a copy of this title";
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: This message refers to these as title level requests. In general, are we referring to these as title or instance level requests?

We use instance in the path and  instanceId as an input parameter, I don’t know if we use instance in other messages or in the discovery system presentation (I imagine that is title).

I’m aware there is a tension in the language in this area, and that might not be solvable here, or the PO may have provided this message. Just thought I’d ask.

Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: I will double check with the PO on messaging.

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 3]
```diff
@@ -72,6 +75,7 @@
 
   private final Logger log;
   private static final String ITEM_ID_FIELD = "itemId";
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Maybe use `RequestProperties.ITEM_ID` instead of defining one here (I realize this code is not part of the PR, but just noticed it).

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 4]
```diff
@@ -109,31 +113,25 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
...
-                     .thenAccept(result -> result.writeTo(routingContext.response()))
-                     .exceptionally( err -> {
-                         String reason = "Error processing title-level request";
-                         log.error(reason, err);
-                         ServerErrorResponse.internalError(routingContext.response(), reason);
-                         return null;
-                      });
+    finder.getItemsByInstanceId(requestByInstanceIdRequest.getInstanceId())
+          .thenCompose(r -> r.after(items -> getRequestQueues(items,requestRelatedRecords, clients)))
+          .thenApply(r -> r.next(items -> validateRequester(items, requestByInstanceIdRequest.getRequesterId())))
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: I guess the downside here is we are now doing everything in order. That's a shame. 😞

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 5]
```diff
@@ -109,31 +113,25 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
...
-                     .thenApply(CreatedJsonResponseResult::from)
-                     .thenAccept(result -> result.writeTo(routingContext.response()))
-                     .exceptionally( err -> {
-                         String reason = "Error processing title-level request";
-                         log.error(reason, err);
-                         ServerErrorResponse.internalError(routingContext.response(), reason);
-                         return null;
-                      });
+    finder.getItemsByInstanceId(requestByInstanceIdRequest.getInstanceId())
+          .thenCompose(r -> r.after(items -> getRequestQueues(items,requestRelatedRecords, clients)))
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]:
```suggestion
          .thenCompose(r -> r.after(items -> getRequestQueues(items, requestRelatedRecords, clients)))
```

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 6]
```diff
@@ -209,11 +202,12 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
           && (instanceRequestPackage.getSortedAvailableItems() == null || instanceRequestPackage.getSortedAvailableItems().isEmpty())) {
           //fail the requests when there are no items to make requests from.
           log.error("Failed to find request queues for all items of instanceId {}",
-            unsortedUnavailableItems.get(0).getInstanceId());
+            items.toArray(new Item[0])[0].getInstanceId());
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: I know this is just an error message, but this code makes me sad. 😄 How about:
```suggestion
            items.iterator().next().getInstanceId());
```
Or better yet, can't we just get the instance id from the `instanceRequestPackage`:
```suggestion
           instanceRequestPackage.getInstanceLevelRequest().getInstanceId().toString());
```

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 7]
```diff
@@ -72,6 +75,7 @@
 
   private final Logger log;
   private static final String ITEM_ID_FIELD = "itemId";
+  private static final String REQUESTER_ID_FIELD = "requesterId";
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Maybe move this to `RequestProperties`?

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 8]
```diff
@@ -109,31 +113,25 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
...
+          .thenCompose(r -> r.after(relatedRecords -> getLoanItems(relatedRecords, clients)))
+          .thenApply( r -> r.next(loanItems -> combineWithUnavailableItems(loanItems, requestRelatedRecords)))
+          .thenApply( r -> r.next(RequestByInstanceIdResource::instanceToItemRequests))
+          .thenCompose( r -> r.after( requests -> placeRequests(requests, clients)))
+          .thenApply(r -> r.map(RequestAndRelatedRecords::getRequest))
+          .thenApply(r -> r.map(new RequestRepresentation()::extendedRepresentation))
+          .thenApply(CreatedJsonResponseResult::from)
+          .thenAccept(result -> result.writeTo(routingContext.response()))
+          .exceptionally( err -> {
+             String reason = "Error processing title-level request";
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: When we come to consensus on title vs instance, we should update this message as well. I believe we should stick to FOLIO terms, instance in this case, but will leave the decision to the PO.

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 1 above)
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: If we still want to use the itemId here, can we change this to `matchingRequest.get().getItemId()` instead? This will save us from doing the `copy()` in `asJson()`.

File: src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java [hunk 9]
```diff
@@ -14,6 +15,7 @@
   private List<Item> itemsWithoutLoans;
   private List<Item> itemsWithoutRequests;
   private RequestByInstanceIdRequest instanceLevelRequest;
+  private Map<Item, RequestQueue> itemRequestQueueMap;
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java]: Should this be part of `Item` someday?

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 1 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: We can make this happen.

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 10]
```diff
@@ -364,33 +358,63 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
...
   }
+
+  private Result<Collection<Item>> validateRequester(Map<Item, RequestQueue> itemRequestQueueMap, UUID requesterId) {
+    if (!itemRequestQueueMap.isEmpty()) {
+      Collection<RequestQueue> requestQueues = itemRequestQueueMap.values();
+      for (RequestQueue queue : requestQueues) {
+        final Optional<Request> matchingRequest = queue.getRequests()
+          .stream()
+          .filter(request -> request.isOpen()
+                           && Objects.equals(request.asJson().getString(REQUESTER_ID_FIELD),
```
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: good question. The Request already had a property to get the Requester that we could use, but in my tests it's always NULL. I wasn't sure if that is due to the test fixture setup or if in real life this method may or may not return NULL, so I erred on the safe side by going to the JSON directly.  I can fix the test fixture if necessary.  @marc-johnson, do you think it's a test-setup issue?

[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Actually, `Request.getUserId()` returns the `requesterId`.

[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: As far as `getRequester()` being `null`. I could see that in a test, but the `requesterId` should never be. I didn't notice `getUserId()` already returned the `requesterId` or I'd just have suggested using that instead.

[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: I'll use getUserId()

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 3 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Good point.

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 2 above)
Comment (magdazacharska) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: We refer to those requests as Title Level Requests.  I also think that "title" is more meaningful to patrons than "instance".  We could compromise on: "This requester already has an open request for a copy of this title (instance)"

Comment (mreno-EBSCO) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Maybe we can make these FOLIO consistent once this is in place: https://issues.folio.org/browse/UXPROD-266

"copy" vs "item" is another case.

File: src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java (hunk 9 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java]: I thought @marc-johnson says it's not that useful to have request queues as part of item, referring to a comment on a PR a few weeks ago when I attempted but failed to attach request queues to items due to circular references.

Comment (mreno-EBSCO) [reply src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java]: OK. Didn't recall that. It just seems naturally part of an `Item` to me, at least in the circulation sense. :)

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java]: I think mod-circulation has a challenge in its domain modelling where too many parts of a entity can be null or only relevant in certain contexts. 

I think the entities may need shrinking rather than growing.

Making every item have to include a request queue suggests that the request queue is relevant everywhere an item is relevant. Maybe that’s the case, and every time we use an item we should get the request queue for it.

Comment (mtraneis) [reply src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java]: If it's relevant to have request queues in an item then perhaps we can make them lazy loaded to not performance worse.

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 4 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Unfortunately

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 6 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: indeed. Used  instanceRequestPackage.getInstanceLevelRequest().getInstanceId()

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 7 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Yes

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: I imagine this is because not all places that use requests fetch the requesting user. The queue is probably one of those places.

The request in a queue should probably be different to a request being used specifically.

File: src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java (hunk 9 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java]: We could potentially add lady loading. That would mean an item either needs wrapping in a transparent proxy (like hibernate does) or an item would need to know how to fetch a queue

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/InstanceRequestRelatedRecords.java]: At some point I’m going to experiment with smaller aggregates in the domain

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 2 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Magda, Matt, and I discussed about this at length and Magda decided to go with the FOLIO terminology "instance" and "item" for now, but will create a mapping for messages in EDS in the near future for a more friendly message.
